jps-plugin/src/org/intellij/erlang/jps/builder/ErlangBuilderUtil.java

```diff
@@ -0,0 +1,69 @@
+package org.intellij.erlang.jps.builder;
+
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.util.JDOMUtil;
+import com.intellij.util.SystemProperties;
+import com.intellij.util.xmlb.SkipDefaultValuesSerializationFilters;
+import com.intellij.util.xmlb.XmlSerializer;
+import org.jdom.Document;
+import org.jdom.JDOMException;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.jps.incremental.CompileContext;
+
+import java.io.File;
+import java.io.IOException;
+
+
+public class ErlangBuilderUtil {
+  public static final String BUILDER_DIRECTORY = "erlang-builder";
+  public static final String ORDERED_DIRTY_FILES_PREFIX = "dirty-files-";
+  public static final String BUILD_ORDER_FILE_NAME = "deps-tree.xml";
+
+  static final Logger LOG = Logger.getInstance(ErlangBuilder.class);
+
+  private ErlangBuilderUtil() {
+  }
+
+  @Nullable
+  public static <T> T readFromXML(@NotNull String filename,
+                                  @NotNull CompileContext context,
+                                  @NotNull Class<T> tClass) {
+    try {
+      File xmlFile = getXmlFile(filename, context);
+      if (!xmlFile.exists()) return null;
+
+      Document document = JDOMUtil.loadDocument(xmlFile);
+      return XmlSerializer.deserialize(document, tClass);
+    }
+    catch (JDOMException e) {
+      LOG.error("Can't read XML from " + filename, e);
+    }
+    catch (IOException e) {
+      LOG.warn("Can't read " + filename, e);
+    }
+    return null;
+  }
+
+  public static void writeToXML(@NotNull String relativePath,
+                                @NotNull ErlangModuleBuildOrder serializedObject,
+                                @NotNull CompileContext context) {
+    try {
+      Document document = new Document(XmlSerializer.serialize(serializedObject, new SkipDefaultValuesSerializationFilters()));
+      File depsConfigFile = getXmlFile(relativePath, context);
+      //noinspection ResultOfMethodCallIgnored
+      depsConfigFile.getParentFile().mkdirs();
+      JDOMUtil.writeDocument(document, depsConfigFile, SystemProperties.getLineSeparator());
+    }
+    catch (IOException e) {
+      LOG.warn("Can't write " + relativePath, e);
+    }
+  }
+
+  @NotNull
+  public static File getXmlFile(@NotNull String filename, @NotNull CompileContext context) {
+    File dataStorageRoot = context.getProjectDescriptor().dataManager.getDataPaths().getDataStorageRoot();
+    File parentDirectory = new File(dataStorageRoot, BUILDER_DIRECTORY);
+    return new File(parentDirectory, filename);
+  }
```

Comment:
Look at these three methods. Each of them has this 'context' parameter, and a 'filename' parameter, yet in each method these arguments are ordered differently.

For cases like this, as a rule of thumb, we first pass a context (usually it is an instance of Project, or some other 'context' object), and then other arguments. It just becomes more structured and easier to read and use.